### PR TITLE
TEST/EXAMPLES: Fix AM Zcopy proto usage in uct_hello_world

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

Register local memory handle only if `md_attr.md_attr.cap.flags` requires it (i.e. `UCT_MD_FLAG_NEED_MEMH` set)

## Why ?

To fix error in `uct_hello_world -z ...` with TCP transport

## How ?

1. Query MD attributes.
2. Check whether we should register a communication buffer to get local memory handle.
3. Do AM Zcopy send.

### CODE REVIEW

File: test/examples/uct_hello_world.c [hunk 1]
```diff
@@ -156,12 +157,18 @@ void zcopy_completion_cb(uct_completion_t *self, ucs_status_t status)
...
                          const cmd_args_t *cmd_args, char *buf)
 {
+    ucs_status_t status = UCS_OK;
     uct_mem_h memh;
     uct_iov_t iov;
     zcopy_comp_t comp;
 
-    ucs_status_t status = uct_md_mem_reg(if_info->md, buf, cmd_args->test_strlen,
-                                         UCT_MD_MEM_ACCESS_RMA, &memh);
+    if ((if_info->md_attr.cap.flags & UCT_MD_FLAG_NEED_MEMH)) {
```
Comment (yosefe) [inline test/examples/uct_hello_world.c]: redundant ( )

File: test/examples/uct_hello_world.c [hunk 2]
```diff
@@ -156,12 +157,18 @@ void zcopy_completion_cb(uct_completion_t *self, ucs_status_t status)
...
+        status = uct_md_mem_reg(if_info->md, buf, cmd_args->test_strlen,
+                                UCT_MD_MEM_ACCESS_RMA, &memh);
+    } else {
+        memh = UCT_MEM_HANDLE_NULL;
```
Comment (yosefe) [inline test/examples/uct_hello_world.c]: what happens when mem_dereg is called?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11031/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 37 workers (click for details)</summary>

**Note:** the logs will be deleted after 13-Jul-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/c336c82037484978947e5a0279037d54f81d031e/_main) | :heavy_check_mark: SUCCESS
[boo31_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/155f92424ad98bd5d0926e780266245f7a25e879/boo31_W0) | :heavy_check_mark: SUCCESS
[boo31_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/c1eee4d5bf1556a58ebd176f4998c6e818c87a1e/boo31_W1) | :heavy_check_mark: SUCCESS
[boo31_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/5d3c46f623c7a90ebc0ac12b0aab863474c001fa/boo31_W2) | :heavy_check_mark: SUCCESS
[boo31_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/4cbe2c1de7fe4730db2b02cc23eaaba885790738/boo31_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/284db10ff9405b85d36c56bd714c1d1110e264f5/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/1a918b8598849003a4718877aa3d22e33f1c40b7/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/5e7288e5adc3caebce681a9db082686cd9e4d78d/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/7f2c5cb30ff243926b2afd26b6a05b5c6c2e2f38/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/fec34e9db1ace25638cb47b9b5e18fe1dc14acc4/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/52d050c602d2fac4d01291456debae99b871e99c/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/6627c8719d7b8b9078ca1a5b054bd4d0d1d1b546/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/dfb07f8d3ecc74b7b64a09bcb27169516f04c3f6/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/081b4b348f8f7623ae21136e5dd71c52805c1b62/hpc-test-node-upstream_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/ebb9e501fbcc44298fa5589dcadeb4533f4d0a43/hpc-test-node-upstream_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/fd5ea3997dc6022b57cbd3226c0eebf6b42578ef/hpc-test-node-upstream_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/ee4c442e70464319f7775e436ba7dc08db920da7/hpc-test-node-upstream_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/b0e3652652a8bb1c7525c5efd2aa314fcd178312/hpc-test-node2_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/88c2b43546e1e7e4a99223200bd9610ac323933a/hpc-test-node2_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/d0e23be8fc77c12910fede8dbc09c37cd9087c08/hpc-test-node2_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/15f27dc44b1b9bfc08717264ea481fb6ea739707/hpc-test-node2_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/8fe7888433f9283682f17ebe87c73fd46fae4dad/hpc-test-node3_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/2eeb9b308cd1dbad81c7c282de86806b35c39e1c/hpc-test-node3_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/88197446bb189b6716c06efca34fe926acd371cf/hpc-test-node3_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/7f5e99c1316179110e63b825cf8deb87878093e0/hpc-test-node3_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/fbaef49b9db2712c8df4bd85d3a8dc5eb57ab331/hpc-test-node_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/4b926bebc5b6caa3db528ccf100e3b943429b285/hpc-test-node_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/4aa54bb8580397f71a609b11be95308d9b3ce113/hpc-test-node_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/25bd50903300cd526e11716ca21c21b9f8d83a46/hpc-test-node_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/308696f4355035e91a03c119ead9fb446e48e858/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/ea3567121e3043766aedc7631d960e4a3640ac2e/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/ef1add29c74b6e0161b9d7ea39678199f0a6d239/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/c80a3572ad6ddbde4e01335ca0f698edb44c7c0f/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/c0ec61945feac3504b8c7c0fbbf6333026520317/vulcan-jenkins_W0) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/98178b2b7ebe444f0606c653cadd6b2769fa3f64/vulcan-jenkins_W1) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/c927ee1a83fc609f8b42feb61517026224f3c3cb/vulcan-jenkins_W2) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/fd0a2f70030bf8010bfc2badbc700c0d/raw/89935c0eeedb257e121d8f2fc2d9592e3037024e/vulcan-jenkins_W3) | :heavy_check_mark: SUCCESS

File: test/examples/uct_hello_world.c (hunk 1 above)
Comment (dmitrygladkov) [reply test/examples/uct_hello_world.c]: fixed

File: test/examples/uct_hello_world.c (hunk 2 above)
Comment (dmitrygladkov) [reply test/examples/uct_hello_world.c]: yes, need to handle this
for TCP it was OK, because `uct_ep_am_zcopy` returns `UCS_OK`
thank you

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/11037/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 37 workers (click for details)</summary>

**Note:** the logs will be deleted after 13-Jul-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/a19e58a3062c1db55b3cdea79800c9030aa983f9/_main) | :heavy_check_mark: SUCCESS
[boo31_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/7a4675e4ceb9599122ee66f6b5c256fe914ead89/boo31_W0) | :heavy_check_mark: SUCCESS
[boo31_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/eb4554c8870cc977e9b5c49a0eee143aff28bfe5/boo31_W1) | :heavy_check_mark: SUCCESS
[boo31_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/c412533ff321282905a7e27899102a4e27e36a0c/boo31_W2) | :heavy_check_mark: SUCCESS
[boo31_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/232067f2bb741c1d398812c7233000a3f944db94/boo31_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/d2a6c5b8b1e9fa736bbf00e1c3212ade8646fcc6/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/32632731e3980880adbe4a28122c932d1f61bdc8/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/a6c5e0d5afe9e3fa9619668057b9bf16c0b3f921/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/878e0e16fa6ea48aa5f3f0b012d6c1a005ac5fbf/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/7a8ddf2fdb69c2726250fe0816a7f7ebecfee164/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/60d1dc3ea4e442179ad706956df81a02a28d2cb1/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/e4dbce9953e6740670ad438601387abd86035abf/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/2351714ef29d876010c373d45335f43431de75d2/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/6c9df4b9882dad4d6e2af17593b656b5e9bdf1c2/hpc-test-node-upstream_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/6883ada7c8c9a39248f43ecb5685a0c419c77e89/hpc-test-node-upstream_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/22927440fae40be1a0c6707f933525e3b4a357ae/hpc-test-node-upstream_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/72a74d2ab3ac45df7ba76d072d793359cb299f62/hpc-test-node-upstream_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/947d0dbf18f52b9eaaa090fd568f71a7a852cc94/hpc-test-node2_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/14b47c3a4a57b49f7e6e0c409778301db0751aec/hpc-test-node2_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/76d110651aa61f4cd0906086b787a221a86cc235/hpc-test-node2_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/a8ddf6c2226a11b608992efa4f6a4f407f9b8964/hpc-test-node2_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/4005f68567e14487d059e886a1594be348cd705e/hpc-test-node3_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/a7208b9bc24db4732622c47f0e0fadf71b713727/hpc-test-node3_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/154334bb53b661de6b100c5354af695621c742e1/hpc-test-node3_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/5a70c41344ee46adb8a36e7dd1d05c6e5947fad0/hpc-test-node3_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/29663e08559d79907d5484b8b7503bf928b019a9/hpc-test-node_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/c46d9aa1e6c098a50a23a0ca98e8bc783d1fe050/hpc-test-node_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/66268299e1f1bbe9bee6b46998d5492bf61b55ff/hpc-test-node_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/90c088fe0fec7445cae83f8bd4464f4fe7cae3f3/hpc-test-node_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/0e5f735467b60ec9da6a62cd48fd7cd8ba68d5a5/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/b9dbeb2a132dbfeb73bdfccf79fa49b15a62aaf9/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/0dad208f77ceb25a10aba99d2d7836bfcd6576a4/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/d2c0ba707f6149ad01f17fc03b7dc71f58d3dc97/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/1e079a9acec25c025edc3928a2a2bc5eddb856ad/vulcan-jenkins_W0) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/76afb6b8250ee0b3b0ead5dc0b36fdf914c67787/vulcan-jenkins_W1) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/12e6aad16f9db9c62e5f5c72921d853fb96880a3/vulcan-jenkins_W2) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/c241ce1c8a90772729c6f2f1420c861c/raw/6fbeb193ad75c7db139e99b2f0a5093d322e0065/vulcan-jenkins_W3) | :heavy_check_mark: SUCCESS

**Pull request merged**